services/ui_backend_service/api/search.py

```diff
@@ -0,0 +1,268 @@
+            # Prefer run_id over run_number and task_name over task_id
+            pathspecs = ["{flow_id}/{run_id}/{step_name}/{task_name}/{name}/{attempt_id}".format(
+                flow_id=art['flow_id'],
+                run_id=art.get('run_id') or art['run_number'],
+                step_name=art['step_name'],
+                task_name=art.get('task_name') or art['task_id'],
+                name=art['name'],
+                attempt_id=art['attempt_id']) for art in meta_artifacts]
+            res = await self._artifact_store.cache.SearchArtifacts(
+                pathspecs, value, operator,
+                invalidate_cache=invalidate_cache)
+
+            if res.has_pending_request():
+                async for event in res.stream():
+                    await ws.send_str(json.dumps(event))
+                await res.wait()
+            artifact_data = res.get()
+
+            results = await _search_dict_filter(meta_artifacts, artifact_data)
+        self.union_results(results)
+
+        await ws.send_str(json.dumps({"event": search_result_event_msg(self.search_results)}))
+
+    async def get_run_artifacts(self, flow_name, run_key, artifact_name):
+        """
+        Find a set of artifacts to perform the search over.
+        Includes localstore artifacts as well, as we want to return that these could not be searched over.
+        """
+        run_id_key, run_id_value = translate_run_key(run_key)
+        db_response = await self._artifact_table.get_records(
+            filter_dict={
+                "flow_id": flow_name,
+                run_id_key: run_id_value,
+                "name": artifact_name
+            }
+        )
+        db_response = await apply_run_tags_to_db_response(flow_name, run_key, self._run_table, db_response)
+        return db_response.body
+
+    async def get_run_metadata(self, flow_name, run_key, metadata_name):
+        """
+        Find a set of artifacts to perform the search over.
+        Includes localstore artifacts as well, as we want to return that these could not be searched over.
+        """
+        run_id_key, run_id_value = translate_run_key(run_key)
+        db_response = await self._metadata_table.get_records(
+            filter_dict={
+                "flow_id": flow_name,
+                run_id_key: run_id_value,
+                "field_name": metadata_name
+            }
+        )
+        return db_response.body
+
+    def union_results(self, new_results: dict):
+        """
+        Union new results with existing results, and update the keyset to prevent duplicate results.
+        Key is determined by step_name and task_id.
+        """
+        for result in new_results:
+            key = result['step_name'] + '/' + result['task_id']
+            if key in self.search_result_keyset:
+                continue
+
+            self.search_result_keyset.add(key)
+            self.search_results.append(result)
```

Comment:
Thanks for pointing this out. I updated the logic here and added a request identifier (with flow/run id + search term). This should reduce the chance of collision (if we still see problem, we can then consider adding timestamp to the identifier keystring but it seems to be overkill for now).

Reply:
yep, chance of collision would seem to be reduced to users competing on the same run with the same search term now, which is very unlikely I would imagine. I still came up with two other approaches that completely avoid the issue if you have time to check those out for viability. Otherwise we can proceed with this and refine it later as well.